### DISCUSSION TITLE
How to fix the stages of the backbone when I finetune a model?

### DISCUSSION THREAD

**Discussion by ZJU-lishuang** [open-mmlab/mmaction2#650] [Q&A] [answered]

**Comment by congee524** [comment]
You can reference https://github.com/open-mmlab/mmaction2/blob/0149a0e8c1e0380955db61680c0006626fd008e9/mmaction/models/backbones/x3d.py#L458

**Comment by ZJU-lishuang** [comment]
I try to modify the config.
`python
model = dict(backbone=dict(frozen_stages=2))
`

single gpu is ok.
But multiple gpus meet error.
`
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by (1) passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`; (2) making sure all `forward` function outputs participate in calculating loss. If you already have done the above two steps, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue(e.g. list, dict, iterable).
`

**Comment by innerlee** [comment]
@ZJU-lishuang Which backbone are you using? The backbone codes may need to be modified too. @congee524 could offer some help if more details are provided.

**Comment by congee524** [comment]
> I try to modify the config.
> `python model = dict(backbone=dict(frozen_stages=2)) `
> 
> single gpu is ok.
> But multiple gpus meet error.
> `RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by (1) passing the keyword argument`find_unused_parameters=True`to`torch.nn.parallel.DistributedDataParallel`; (2) making sure all `forward`function outputs participate in calculating loss. If you already have done the above two steps, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's`forward`function. Please include the loss function and the structure of the return value of`forward`of your module when reporting this issue(e.g. list, dict, iterable).`

Try to set `find_unused_parameters = True` in your config file.

**Comment by ZJU-lishuang** [comment, marked as answer]
shell
`
bash tools/dist_train.sh configs/recognition/tsm/tsm_r50_1x1x8_50e_realviolent_rgb.py 2 --work-dir work_dirs/tsm_r50_1x1x8_50e_realviolent_lr0025_k400_freeze --validate --seed 0 --deterministic
`
config

[tsm_r50_1x1x8_50e_realviolent_rgb.txt](https://github.com/open-mmlab/mmaction2/files/6041077/tsm_r50_1x1x8_50e_realviolent_rgb.txt)

And using `find_unused_parameters = True` solves the problem.
Thanks.

**Comment by irvingzhang0512** [comment]
@innerlee @congee524 How about adding some docs for all `frozen_stages`, i.e. [here](https://github.com/open-mmlab/mmaction2/blob/5c1093e20efeb4c931404b51701ca9552facacfb/mmaction/models/backbones/resnet.py#L310), remind users to set `find_unused_parameters = True` for distributed training/testing

**Comment by innerlee** [comment]
Add it to FAQ
